src/features/optional-chaining.md

```diff
@@ -0,0 +1,138 @@
+// only for admin users.
+const adminOption = db?.user?.validateAdminAndGetPrefs?.().option;
+```
+
+The syntax can feel unexpected, as the `?.()` is the actual operator, which applies to the expression _before_ it.
+
+There’s a third usage of the operator, namely the optional dynamic property access, which is done via `?.[]`. It either returns the value referenced by the argument in the brackets, or `undefined` if there’s no object to get the value from. Here’s a possible use case, following the example from above:
+
+```js
+// Extends the capabilities of the static property access
+// with a dynamically generated property name.
+const optionName = 'optional setting';
+const optionLength = db?.user?.preferences?.[optionName].length;
+```
+
+This last form is also available for optionally indexing arrays, e.g.:
+
+```js
+// If the `usersArray` is `null` or `undefined`,
+// then `userName` gracefully evaluates to `undefined`.
+const userIndex = 42;
+const userName = usersArray?.[userIndex].name;
+```
+
+The optional chaining operator can be combined with the [nullish coalescing operator](https://github.com/tc39/proposal-nullish-coalescing) when a non-`undefined` default value is needed. This enables safe deep property access with a specified default value, addressing a common use case that previously required userland libraries such as [lodash’s `_.get`](https://lodash.dev/docs/4.17.15#get):
+
+```js
+const object = { id: 123, names: { first: 'Alice', last: 'Smith' }};
+
+{ // With lodash:
+  const firstName = _.get(object, 'names.first');
+  // → 'Alice'
+
+  const middleName =_.get(object, 'names.middle', '(no middle name)');
+  // → '(no middle name)'
+}
+
+{ // With optional chaining and nullish coalescing:
+  const firstName = object?.names?.first ?? '(no first name)';
+  // → 'Alice'
+
+  const middleName = object?.names?.middle ?? '(no middle name)';
+  // → '(no middle name)'
+}
+```
+
+## Properties of the optional chaining operator { #properties }
+
+The optional chaining operator has a few interesting properties: _short-circuiting_, _stacking_, _grouping_, and _optional deletion_. Let's walk through each of these with an example.
+
+_Short-circuiting_ means not evaluating the rest of the expression if an optional chaining operator returns early:
+
+```js
+// `age` is incremented only if `db` and `user` are defined.
+db?.user?.grow(++age);
+```
+
+_Stacking_ means that more than one optional chaining operator can be applied on a sequence of property accesses:
+
+```js
+// An optional chain may be followed by another optional chain.
+const firstNameLength = db?.users[42].names?.first.length;
+```
+
+Still, be considerate about using more than one optional chaining operator in a single chain. If a value is guaranteed to not be nullish, then using `?.` to access properties on it should be discouraged. For instance in the example above, `users` and `users[42]` are always considered to be defined.
```

Comment:
Thanks for adding the new section! I think we can tweak the example a little bit more:

```js
const firstNameLength = db?.users[42].names?.first.length;
```

In practice, it's probably more common for the root object itself to always be defined, but then any properties on it to not be. How about, instead of the above, something like:

```js
const firstNameLength = db.users?.[42].names?.first.length;
```

Then you could say:

> In the example above, `db` is considered to always be defined, but `db.users` and `db.users[42]` may not be. If the user has a `names` property, then `names.first.length` is assumed to always be defined.

WDYT?

Reply:
Thanks, PTAL at the new version.